core/modules/loader/BufferUdp.cc

```diff
@@ -0,0 +1,158 @@
+        _wCursor += len; /// must advance the cursor.
+
+        // EOF is only a problem if no MsgElement was retrieved.
+        // ??? This is definitely the case in UDP, EOF as nothing more will show up.
+        // ??? But TCP is another issue as EOF is returned when the connection is still live but
+        // ??? there's no data (len=0). Why does read_some set error to eof before the tcp connection is closed?
+        if (error == boost::asio::error::eof) {
+            LOGS(_log, LOG_LVL_WARN, "readFromSocket eof");
+            break; // Connection closed cleanly by peer.
+        } else if (error && error != boost::asio::error::eof) {
+            throw LoaderMsgErr("BufferUdp::readFromSocket note=" + note + " asio error=" + error.message());
+        }
+
+        /// Try to retrieve an element (there's no guarantee that an entire element got read in a single read.
+        // Store original cursor positions so they can be restored if the read fails.
+        msgElem = _safeRetrieve();
+        if (msgElem != nullptr) {
+            return msgElem;
+        }
+    }
+    return nullptr;
+}
+
+
+std::shared_ptr<MsgElement> BufferUdp::_safeRetrieve() {
+    auto wCursorOriginal = _wCursor;
+    auto rCursorOriginal = _rCursor;
+    MsgElement::Ptr msgElem = MsgElement::retrieve(*this);
+    if (msgElem != nullptr) {
+        return msgElem;
+    } else {
+        _wCursor = wCursorOriginal;
+        _rCursor = rCursorOriginal;
+    }
+    return nullptr;
+}
+
+
+bool BufferUdp::isRetrieveSafe(size_t len) const {
+    auto newLen = (_rCursor + len);
+    return (newLen <= _end && newLen <= _wCursor);
+}
+
+
+bool BufferUdp::retrieve(void* out, size_t len) {
+    if (isRetrieveSafe(len)) {
+        memcpy(out, _rCursor, len);
+        _rCursor += len;
+        return true;
+    }
+    return false;
+}
+
+
+bool BufferUdp::retrieveString(std::string& out, size_t len) {
+    if (isRetrieveSafe(len)) {
+        const char* strEnd = _rCursor + len;
+        std::string str(_rCursor, strEnd);
+        _rCursor = strEnd;
+        out = str;
+        return true;
+    }
+    return false;
+}
+
+
+std::string BufferUdp::dumpStr(bool hexDump, bool charDump) const {
+        std::stringstream os;
+        os << "maxLength=" << _length;
+
+        os <<   " buffer=" << (void*)_buffer;
+        os <<  " wCurLen=" << getAvailableWriteLength();
+        os <<  " wCursor=" << (void*)_wCursor;
+        os <<  " rCurLen=" << getBytesLeftToRead();
+        os <<  " rCursor=" << (void*)_rCursor;
+        os <<      " end=" << (void*)_end;
+
+        // hex dump
+        if (hexDump) {
+        os << "(";
+        for (const char* j=_buffer; j < _wCursor; ++j) {
+            os << std::hex << (int)*j << " ";
+        }
+        os << ")";
+        }
+        std::string str(os.str());
+
+        // character dump
+        if (charDump) {
+            str += "(" + std::string(_buffer, _wCursor) + ")";
```

Comment:
Why not to just keep streaming into the output stream? This would be more efficient. Something along the line of:
```
if (charDump) {
    os << "(" << std::string(_buffer, _wCursor) << ")";
}
return os.str();
```

Reply:
it looks like str is returned, not sent to output (but maybe it's streamed later, I don't know...)